core/src/main/scala/kafka/zk/KafkaZkClient.scala

```diff
@@ -87,13 +90,71 @@ class KafkaZkClient private (zooKeeperClient: ZooKeeperClient, isSecure: Boolean
 
   /**
    * Registers a given broker in zookeeper as the controller.
+   * @return the updated controller epoch and epoch zkVersion
    * @param controllerId the id of the broker that is to be registered as the controller.
    * @param timestamp the timestamp of the controller election.
-   * @throws KeeperException if an error is returned by ZooKeeper.
-   */
-  def registerController(controllerId: Int, timestamp: Long): Unit = {
-    val path = ControllerZNode.path
-    checkedEphemeralCreate(path, ControllerZNode.encode(controllerId, timestamp))
+   * @throws ControllerMovedException if fail to create /controller or fail to increment controller epoch.
+   */
+  def registerController(controllerId: Int, timestamp: Long): (Int, Int) = {
+    getControllerEpoch match {
+      case Some(epochAndStat) =>
+        // Create \controller
+        val expectedControllerEpochZkVersion = epochAndStat._2.getVersion
+        info(s"Try to create ${ControllerZNode.path} with expected controller epoch zkVersion $expectedControllerEpochZkVersion")
+        tryCreateControllerZNode(controllerId, timestamp, expectedControllerEpochZkVersion)
+
+        // Update \controller_epoch
+        val curEpoch = epochAndStat._1
+        val newControllerEpoch =
+          if (curEpoch == ControllerEpochZNode.NoEpoch)
+            KafkaController.InitialControllerEpoch
+          else
+            curEpoch + 1
+        info(s"Try to increment controller epoch to $newControllerEpoch with expected controller epoch zkVersion $expectedControllerEpochZkVersion")
+        val setDataResponse = setControllerEpochRaw(newControllerEpoch, expectedControllerEpochZkVersion)
+        setDataResponse.resultCode match {
+          case Code.OK =>
+            (newControllerEpoch, setDataResponse.stat.getVersion)
+          case _ =>
+            throw new ControllerMovedException("Controller moved to another broker. Aborting controller startup procedure")
```

Comment:
Prior to this patch, if setControllerEpochRaw fails and the error is not `NONODE`, ControllerMovedException will be thrown which will be caught and `triggerControllerMove()` will be executed.

After this patch, if setControllerEpochRaw returns a non-OK error code, we will always throw ControllerMovedException(), which will be caught in the upper layer without executing `triggerControllerMove()`. I am wondering if we should throw IllegalStateException if the error code suggests something other than controller move, so that we can still execute `triggerControllerMove()` in this scenario.

Reply:
Per https://github.com/apache/kafka/pull/5101#discussion_r213122229
`/controller_epoch` should exists and `setControllerEpochRaw` should not return `NONODE`. If that does happen, we will rely on admin to recover and `triggerControllerMove()` will not help.
I agree that throwing `ControllerMovedException` is not a good idea. Maybe we should just throw `IllegalStateException` and indicate that is a FATAL error. What do you think?

Reply:
Thanks for the comment. I have updated the PR to throw `ControllerMovedException` only when we see BadVersion in `setControllerEpochRaw`. I also applied the same idea to `maybeCreateControllerZNode`.